### ISSUE TITLE
Refactor > Move to new file -- not working

### ISSUE THREAD

**Issue by babakness** [Microsoft/vscode#58711]
- VSCode Version: 1.27.0 (1.27.0)
- OS Version: MacOS 10.13.5

Steps to Reproduce:

**1. Highlight a function
2. Select refactor > move to new file**

Does this issue occur when all extensions are disabled?: Yes/No

**Yes -- meaning it also fails to work with --disable-extensions**

-- 

I'm not sure why, the feature was working just fine before. I also tried setting `"typescript.tsserver.trace": "verbose"`

Tried on multiple projects

I see nothing in the Output pane.

**Comment by babakness** [comment]
Update: can confirm the feature works on VS Code Version 1.25.1 (1.25.1), does not work on latest.

**Comment by babakness** [comment]
Ah, yeah I see it now. I get this output just my highlighting text (not actually selecting "Move to new file"). I get no output when actually selecting Move To New File. Nothing happens. Works on VS Code 1.25 just fine (my current work around is to use this older build)
```
[Trace  - 3:11:38 PM] Sending request: quickinfo (47). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "...",
    "line": 10,
    "offset": 18
}
[Trace  - 3:11:38 PM] Response received: quickinfo (47). Request took 1 ms. Success: false . Message: No content available.
[Trace  - 3:11:38 PM] Sending request: quickinfo (48). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "...",
    "line": 10,
    "offset": 17
}
[Trace  - 3:11:38 PM] Response received: quickinfo (48). Request took 1 ms. Success: false . Message: No content available.
[Trace  - 3:11:43 PM] Sending request: getApplicableRefactors (49). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "...",
    "startLine": 10,
    "startOffset": 1,
    "endLine": 11,
    "endOffset": 1
}
[Trace  - 3:11:43 PM] Response received: getApplicableRefactors (49). Request took 8 ms. Success: true 
Result: [
    {
        "name": "Convert export",
        "description": "Convert named export to default export",
        "actions": [
            {
                "name": "Convert named export to default export",
                "description": "Convert named export to default export"
            }
        ]
    },
    {
        "name": "Extract Symbol",
        "description": "Extract function",
        "actions": [
            {
                "description": "Extract to function in module scope",
                "name": "function_scope_0"
            }
        ]
    },
    {
        "name": "Extract Symbol",
        "description": "Extract constant",
        "actions": [
            {
                "description": "Extract to constant in enclosing scope",
                "name": "constant_scope_0"
            }
        ]
    },
    {
        "name": "Move to a new file",
        "description": "Move to a new file",
        "actions": [
            {
                "name": "Move to a new file",
                "description": "Move to a new file"
            }
        ]
    }
]
```

**Comment by mjbvz** [comment]
You can switch the typescript version independently of VS Code by following [these instructions](https://code.visualstudio.com/Docs/languages/typescript#_using-newer-typescript-versions). In the latest VS Code insiders build, can you please try:

- Using `typescript@2.9`. This was the version shipped in with VS Code 1.25
- Using `typescript@next`. This has the latest fixes

Do either of those fix the problem?

Let me know if you have any questions about switching the typescript version in your workspace.

**Comment by mjbvz** [comment]
Also, does this happen for every refactoring you try to trigger? Does it happen in a new project if you just try to extract a simple expression like `1 + 2` to a constant or function?

**Comment by babakness** [comment]
Hey, actually I'm using Typescript 3.0.3 on VS Code 1.25 by choosing the "Workspace" Typescript--and refactoring works as expected. 

`1 + 2` refactor does not work in VSC 1.27, does work in VSC 1.25

**Comment by babakness** [comment]
To follow up, every option on the refactoring menu is not working in the latest VS Code, including the "Insiders" edition with no plugins or extensions running.

**Comment by mjbvz** [comment]
Thanks. Unfortunately I'm unable to reproduce this on any machine so I need a bit more information from you:

1. What are the exact steps to take to show the code action menu and the select the desired refactoring? Are you using the mouse? The keyboard?

1. In a TypeScript file, if you write:

```ts
class Foo {}
class Bar extends Foo {
    constructor() {}
}
```
 
Can you successfully trigger the quick fix that adds the missing super call in the constructor?

3. Try collecting VS Code's internal logs for executing the quick fix:

     1. In VS Code, run the command `Developer: Set log level...`
     1. Select `trace`. Restart VS Code and reproduce the problem
     1. Back in the output panel, share the output from the `Log (Extension Host)` and `Log (Window)` sections

**Comment by babakness** [comment]
Hey, so I just noticed something. If I click on the lightbulb icon when I select code, it works! But it doesn't work if I select the code, right click on the selected code, choose "Refactor..." then choose "Move To New File"

Log output

```
[2018-09-19 18:46:36.855] [renderer1] [info] no standard startup: not the explorer viewlet
[2018-09-19 18:46:53.619] [renderer1] [info] no standard startup: panel is active
```

Nothing else!

**Comment by babakness** [comment]
Oh here is potentially useful log, last one was from `Log (window)` this one is from `Log (Extension Host)`

```
[2018-09-19 18:51:37.206] [exthost1] [warning] vscode.typescript-language-features -Code actions of kind 'refactor 'requested but returned code action is of kind 'quickfix'. Code action will be dropped. Please check 'CodeActionContext.only' to only return requested code actions.
```

**Comment by mjbvz** [comment]
Thanks @babakness. I can repo now using the refactoring menu

@sbatten and @bpasero I believe I have tracked this down to the refactor context menu seemingly not firing its actions when the user select an item. This may because the refactor context menu is triggered from within another context menu. I'll keep debugging tomorrow but let me know if you have any ideas on what may be going wrong here

**Comment by bpasero** [comment]
@mjbvz does it reproduce with the custom menu but not the native menu?

**Comment by mjbvz** [comment]
This is on MacOS, so native context menus

**Comment by bpasero** [comment]
@mjbvz that would be very weird. I am not even able to bring up anything when I click on "Refactoring...", is this broken? E.g. I selected some code and triggered it and was at least expecting something like "Extract method".

**Comment by bpasero** [comment]
Ok I can reproduce, steps:
* click "Source actions..."
* click "Organize Imports"

=> nothing happens

Works fine when invoked from the command palette.

**Comment by bpasero** [comment]
@mjbvz I verified that the "Organize Imports" action is triggered, so it does not seem to be an issue with the menu.

**Comment by mjbvz** [comment]
This is a bug with how we handle context menus. I added these log points in `vscode/src/vs/base/parts/contextmenu/electron-browser/contextmenu.ts`

<img width="728" alt="screen shot 2018-09-20 at 9 35 12 am" src="https://user-images.githubusercontent.com/12821956/45833055-82d80680-bcb8-11e8-85c7-0dc4f756bf13.png">

which shows the following sequence of events.

<img width="410" alt="screen shot 2018-09-20 at 9 30 30 am" src="https://user-images.githubusercontent.com/12821956/45832848-e6156900-bcb7-11e8-9cca-36a265d0d073.png">

To summarize:

1. Open the editor context menu
1. Select the refactoring option
1. Which opens another context menu
1. At this point we receive a generic `context menu closed` event.
1. This event removes both the original editor context menu's click handler and the click handler for the refactor context menu we just created

This explains why adding a slight delay when creating the refactor context menu fixes the problem

**Comment by bpasero** [comment]
> This event removes both the original editor context menu's click handler and the click handler for the refactor context menu we just created

Where?

**Comment by mjbvz** [comment]
The editor and refactor menus both register a close event handler. When the editor context menu is closed, both of the menus receive the same close event so we end up removing the refactor menu’s click handler too even though is only just started showing
